File: netlify/functions/analyze.py
import json, subprocess, sys, os
from pathlib import Path

COOKIES_MAP = {
    'youtube': 'youtube-cookies.txt',
    'tiktok': 'tiktok-cookies.txt',
    'instagram': 'instagram-cookies.txt',
    'pinterest': 'pinterest-cookies.txt',
    'facebook': 'facebook-cookies.txt'
}

def cookies_path(platform):
    base = Path(__file__).parent / 'cookies'
    return base / COOKIES_MAP.get(platform, 'youtube-cookies.txt')

def handler(event, context):
    if event.get('httpMethod') != 'POST':
        return {'statusCode': 405, 'body': 'Method Not Allowed'}

    try:
        body = json.loads(event.get('body', '{}'))
        url = body.get('url', '').strip()
        platform = body.get('platform', 'youtube').lower()
        if not url:
            return {'statusCode': 400, 'body': json.dumps({'error': 'URL required'})}

        cmd = [
            sys.executable, '-m', 'yt_dlp',
            '--dump-single-json',
            '--no-warnings',
            '--no-call-home',
            '--cookies', str(cookies_path(platform)),
            url
        ]
        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
        info = json.loads(result.stdout)

        safe = {
            'id': info.get('id'),
            'title': info.get('title'),
            'extractor': info.get('extractor'),
            'uploader': info.get('uploader'),
            'duration': info.get('duration'),
            'duration_string': info.get('duration_string'),
            'thumbnail': info.get('thumbnail'),
            'webpage_url': info.get('webpage_url'),
            'formats': [{
                'format_id': f.get('format_id'),
                'ext': f.get('ext'),
                'height': f.get('height'),
                'width': f.get('width'),
                'fps': f.get('fps'),
                'acodec': f.get('acodec'),
                'vcodec': f.get('vcodec'),
                'filesize': f.get('filesize'),
                'filesize_approx': f.get('filesize_approx'),
                'format_note': f.get('format_note')
            } for f in info.get('formats', [])],
            'entries': [{
                'id': e.get('id'),
                'title': e.get('title'),
                'thumbnail': e.get('thumbnail'),
                'url': e.get('url'),
                'webpage_url': e.get('webpage_url'),
                'original_url': e.get('original_url')
            } for e in info.get('entries', [])[:60]]
        }
        return {
            'statusCode': 200,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps(safe)
        }
    except subprocess.CalledProcessError as e:
        return {'statusCode': 500, 'body': json.dumps({'error': e.stderr})}
    except Exception as e:
        return {'statusCode': 500, 'body': json.dumps({'error': str(e)})}

File: netlify/functions/analyze.js
const { spawn } = require('child_process');
const path = require('path');

const PLATFORM_COOKIES = {
  youtube: 'youtube-cookies.txt',
  tiktok: 'tiktok-cookies.txt',
  instagram: 'instagram-cookies.txt',
  pinterest: 'pinterest-cookies.txt',
  facebook: 'facebook-cookies.txt'
};

function cookiesPath(platform) {
  return path.join(process.cwd(), 'cookies', PLATFORM_COOKIES[platform] || 'youtube-cookies.txt');
}

function runYtDlp(args) {
  return new Promise((resolve, reject) => {
    const child = spawn('python3', ['-m', 'yt_dlp', ...args], {
      cwd: process.cwd(),
      env: { ...process.env, PYTHONUNBUFFERED: '1' }
    });
    let stdout = '';
    let stderr = '';
    child.stdout.on('data', (chunk) => { stdout += chunk.toString(); });
    child.stderr.on('data', (chunk) => { stderr += chunk.toString(); });
    child.on('error', reject);
    child.on('close', (code) => {
      if (code === 0) resolve(stdout);
      else reject(new Error(stderr || `yt-dlp exited with code ${code}`));
    });
  });
}

exports.handler = async (event) => {
  if (event.httpMethod !== 'POST') {
    return { statusCode: 405, body: JSON.stringify({ error: 'Method not allowed' }) };
  }

  try {
    const body = JSON.parse(event.body || '{}');
    const url = String(body.url || '').trim();
    const platform = String(body.platform || 'youtube').toLowerCase();
    if (!url) return { statusCode: 400, body: JSON.stringify({ error: 'URL is required' }) };

    const args = [
      '--dump-single-json',
      '--no-warnings',
      '--no-call-home',
      '--cookies', cookiesPath(platform),
      url
    ];

    const output = await runYtDlp(args);
    const info = JSON.parse(output);

    const safe = {
      id: info.id,
      title: info.title,
      extractor: info.extractor,
      uploader: info.uploader,
      duration: info.duration,
      duration_string: info.duration_string,
      thumbnail: info.thumbnail,
      webpage_url: info.webpage_url,
      formats: (info.formats || []).map((format) => ({
        format_id: format.format_id,
        ext: format.ext,
        height: format.height,
        width: format.width,
        fps: format.fps,
        acodec: format.acodec,
        vcodec: format.vcodec,
        filesize: format.filesize,
        filesize_approx: format.filesize_approx,
        format_note: format.format_note
      })),
      entries: Array.isArray(info.entries) ? info.entries.slice(0, 60).map((entry) => ({
        id: entry.id,
        title: entry.title,
        thumbnail: entry.thumbnail,
        url: entry.url,
        webpage_url: entry.webpage_url,
        original_url: entry.original_url
      })) : []
    };

    return {
      statusCode: 200,
      headers: { 'Content-Type': 'application/json', 'Cache-Control': 'no-store' },
      body: JSON.stringify(safe)
    };
  } catch (error) {
    return {
      statusCode: 500,
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ error: error.message })
    };
  }
};
